sw/host/opentitanlib/src/otp/vmem_serialize.rs

```diff
@@ -0,0 +1,278 @@
+        let size = bytes.len();
+        VmemItem {
+            bytes: Some(bytes),
+            size,
+            offset,
+            name,
+        }
+    }
+
+    pub fn new_unvalued(size: usize, offset: usize, name: String) -> VmemItem {
+        VmemItem {
+            bytes: None,
+            size,
+            offset,
+            name,
+        }
+    }
+}
+
+pub type DigestIV = u64;
+pub type DigestCnst = u128;
+
+/// Digest information for an OTP partition.
+#[derive(PartialEq)]
+pub enum DigestType {
+    Unlocked,
+    Software,
+    Hardware(DigestIV, DigestCnst),
+}
+
+/// The hex representation of an OTP partition.
+pub struct VmemPartition {
+    /// Items associated with this partition.
+    items: Vec<VmemItem>,
+    /// The name of this partition.
+    /// Used in annotations.
+    name: String,
+    /// The type of digest used for this partition.
+    /// For software digests, the value of the digest is provided and appended to the list of
+    /// items. For hardware digests, we must compute the digest value and append to the list of
+    /// items.
+    digest_type: DigestType,
+    /// Partition size.
+    size: usize,
+    /// The key name for this parition.
+    /// If specified, the serializer will attempt to scramble this parition using the key named in
+    /// this field.
+    key_name: Option<String>,
+}
+
+impl VmemPartition {
+    pub fn new(
+        name: String,
+        size: usize,
+        digest_type: DigestType,
+        key_name: Option<String>,
+    ) -> VmemPartition {
+        VmemPartition {
+            items: Vec::new(),
+            name,
+            digest_type,
+            size,
+            key_name,
+        }
+    }
+
+    pub fn adjust_size(&mut self, size: usize) {
+        self.size = size;
+    }
+
+    /// Add an item to this partition.
+    pub fn push_item(&mut self, item: VmemItem) {
+        self.items.push(item);
+    }
+
+    /// Produces a tuple containing OTP HEX lines with annotations.
+    fn write_to_buffer(
+        &self,
+        keys: &HashMap<String, Vec<u8>>,
+    ) -> Result<(Vec<u8>, Vec<String>)> {
+        if self.size % 8 != 0 {
+            return Err(anyhow!("Partition must be 64-bit alligned."));
+        }
+
+        let mut defined = vec![false; self.size];
+        let mut annotations: Vec<String> = vec!["unallocated".to_owned(); self.size];
+
+        let mut data_bytes: Vec<u8> = vec![0; self.size];
+
+        for item in &self.items {
+            for i in 0..item.size {
+                let idx = i + item.offset;
+                annotations[idx] = format!("{}: {}", self.name, item.name).to_string();
+                if let Some(bytes) = &item.bytes {
+                    data_bytes[idx] = bytes[i];
+                    if defined[idx] {
+                        return Err(anyhow!("Unexpected item collision at 0x{:x} with item {}", idx, item.name));
+                    }
+                    defined[idx] = true;
+                }
+            }
```

Comment:
Maybe something like this would work better...
```rust
let end = item.offset + item.size;
annotations[item.offset..end].fill(format!(...));
let defined = &mut defined[item.offset..end];
if let Some(collision) = defined.iter().position(|defined| *defined) {
  // ...
}
defined.fill(true);
data_bytes[item.offset..end].copy_from_slice(&item.bytes);
```